### ISSUE TITLE
SampleCmp and SampleCmpLevelZero not working in HLSL.

### ISSUE THREAD

**Issue by johnhable** [KhronosGroup/glslang#698]
SampleCmp() and SampleCmpLevelZero() don't seem to work. As a test, the following shadow code works fine:

float d = g_shadowTex.Sample(g_sampler,shadLookup.xy).x;
float shad = d + 0.0001f < shadLookup.z ? 0.0f : 1.0f;

But doing the same using a depth comparison sampler does not. I.e.:

float shad = g_shadowTex.SampleCmpLevelZero(g_shadowSampler,shadLookup.xy,shadLookup.z).x;

That line just returns true for a sampler with compareEnable set to true and compareOp set to VK_COMPARE_OP_LESS_OR_EQUAL. ALWAYS and NEVER work correctly, and SampleCmpLevelZero returns 1.0 and 0.0 respectively. But the less than op always returns 1.0. In fact, both of these snippets return the same value (1.0):

float shad = g_shadowTex.SampleCmp(g_shadowSampler,shadLookup.xy,-1000.0);
float shad = g_shadowTex.SampleCmp(g_shadowSampler,shadLookup.xy,1000.0);

My intuition would be that the Z component isn't getting set to the right place in the spirv code.

**Comment by steve-lunarg** [comment]
Hi @johnhable,

This is a good one :-)

I've been staring at the code we produce for your SampleCmpLevelZero example, and so far I've yet to spot the problem.  There are some sequences that are hard to follow though, and I might be overlooking something still, so I will keep digging.

Out of curiosity, have you seen this behavior on multiple vendor's implementations?  I am wondering in particular due to this spec language:

> Coordinate must be a scalar or vector of floating-point type. ...  It may be a vector larger than needed, but all unused components will appear after all used components.

Indeed, the SPIR-V being generated does supply a vector that's one component larger than needed.  I could imagine room for defects, although on the other hand, I would image this would be already shaken out through the GLSL path, so I'm still favoring the theory that I'm simply overlooking the bug.

**Comment by johnhable** [comment]
Hi Steve,

It's only been tested on an NVIDIA machine, because that's what I have. But that extra component is very suspicious. That would seem to explain the behavior that I'm seeing (ALWAYS is always 1, NEVER is always 0, and LESS is always 0 regardless of the number I'm putting in). Has this path been verified on an AMD card?

John

**Comment by johnhable** [comment]
Hi,

Ok, we tracked this down. The problem in my specific case was that I was using a SamplerState, instead of a SamplerComparisonState. When using a SamplerState in a SampleCmp() call, the compiler was emitting a OpImageSampleImplicitLod() instead of OpImageSampleDrefImplicitLod().

So the generated code is fine, but we should probably make it return an error in this case as opposed to silently changing the SampleCmp() into a Sample() call. I.e. this code should return an error:

Texture2D g_shadowTex;
SamplerState g_shadowSampler;
SamplerComparisonState g_shadowSamplerComp;
...
float shad = g_shadowTex.SampleCmp(g_shadowSampler,uv,d); // should be an error
float shad = g_shadowTex.SampleCmp(g_shadowSamplerComp,uv,d); // is fine


John

**Comment by steve-lunarg** [comment]
Cool, glad to hear it's working now!

Agreed about the error.  There are a number of missing error detections in the HLSL FE since it's not currently acting as a validator.  Perhaps we should build a master list of those.

For this specific one, I'll check into it; I believe it should be straightforward to detect it and issue an error.
